devtools/klap-decrypt.py

```diff
@@ -0,0 +1,272 @@
+
+from kasa.credentials import Credentials
+from kasa.deviceconfig import (
+    DeviceConfig,
+    DeviceConnectionParameters,
+    DeviceEncryptionType,
+    DeviceFamily,
+)
+from kasa.klaptransport import KlapEncryptionSession, KlapTransportV2
+
+# These are the only variables that need to be changed in order to run this code.
+# ********** CHANGE THESE **********
+username = "hello@world.com"  # cloud account username (likely an email)
+password = "hunter2"  # cloud account password  # noqa: S105
+# the ip of the smart device as it appears in the pcap file
+device_ip = "192.168.1.100"
+pcap_file_path = "/path/to/my.pcap"  # the path to the pcap file
+output_json_name = (
+    "output.json"  # the name of the output file, relative to the current directory
+)
+# **********************************
+
+capture = pyshark.FileCapture(pcap_file_path, display_filter="http")
+
+# In an effort to keep this code tied into the original code
+# (so that this can hopefully leverage any future codebase updates inheriently),
+# some weird initialization is done here
+myCreds = Credentials(username, password)
+
+fakeConnection = DeviceConnectionParameters(
+    DeviceFamily.SmartTapoBulb, DeviceEncryptionType.Klap
+)
+fakeDevice = DeviceConfig(
+    device_ip, connection_type=fakeConnection, credentials=myCreds
+)
+
+
+# In case any modifications need to be made in the future,
+# this class is created to allow for easy modification
+class MyKlapTransport(KlapTransportV2):
+    """A custom KlapTransportV2 class that allows for easy modification."""
+
+    pass
+
+
+# This is a custom error handler that replaces bad characters with '*',
+# in case something goes wrong in decryption.
+# Without this, the decryption could yield an error.
+def bad_chars_replacement(exception):
+    """Replace bad characters with '*'."""
+    return ("*", exception.start + 1)
+
+
+codecs.register_error("bad_chars_replacement", bad_chars_replacement)
+
+
+class MyEncryptionSession(KlapEncryptionSession):
+    """A custom KlapEncryptionSession class that allows for decryption."""
+
+    def decrypt(self, msg):
+        """Decrypt the data."""
+        decryptor = self._cipher.decryptor()
+        dp = decryptor.update(msg[32:]) + decryptor.finalize()
+        unpadder = padding.PKCS7(128).unpadder()
+        plaintextbytes = unpadder.update(dp) + unpadder.finalize()
+
+        return plaintextbytes.decode("utf-8", "bad_chars_replacement")
+
+
+class Operator:
+    """A class that handles the data decryption, and the encryption session updating."""
+
+    def __init__(self, klap, creds):
+        self.__local_seed: bytes = None
+        self.__remote_seed: bytes = None
+        self.__session: MyEncryptionSession = None
+        self.__creds = creds
+        self.__klap: MyKlapTransport = klap
+        self.__auth_hash = self.__klap.generate_auth_hash(self.__creds)
+        self.__local_auth_hash = None
+        self.__remote_auth_hash = None
+        self.__seq = 0
```

Comment:
I think these should be single underscores for consistency with the library

Reply:
Fixed in 83e512fb65df123f42be3e39557d12cb9ccc338c